packages/react-charting/src/components/SankeyChart/SankeyChart.base.tsx

```diff
@@ -186,57 +204,160 @@ function adjustOnePercentHeightNodes(nodesInColumn: NodesInColumns) {
+ * The links have the original weights. Computed nodes have the total weight of all incoming and outgoing links.
  */
-function populateNodeActualValue(data: ISankeyChartData) {
+function populateNodeActualValue(data: ISankeyChartData, computedNodes: NodeValues, originalLinks: LinkValues) {
   data.links.forEach((link: SLink) => {
     if (!link.unnormalizedValue) {
-      link.unnormalizedValue = link.value;
+      link.unnormalizedValue = linkValue(originalLinks, link);
     }
   });
   data.nodes.forEach((node: SNode) => {
-    node.actualValue = Math.max(
-      d3Sum(node.sourceLinks!, (link: SLink) => link.unnormalizedValue),
-      d3Sum(node.targetLinks!, (link: SLink) => link.unnormalizedValue),
-    );
+    node.actualValue = computedNodes[node.nodeId as number];
+  });
+}
+
+/**
+ * This is used to introduce dynamic padding for cases where the number of nodes in a column is huge
+ * so that we maintain a node to space ratio for such columns as if we fail to do so the
+ * chart is devoid of nodes and only shows links.
+ */
+function adjustPadding(sankey: SankeyLayoutGenerator, height: number, nodesInColumn: NodesInColumns): void {
+  let padding = sankey.nodePadding();
+  const minPadding = PADDING_PERCENTAGE * height;
+  Object.values(nodesInColumn).forEach((column: SNode[]) => {
+    const totalPaddingInColumn = height - d3Sum(column, (node: SNode) => node.y1! - node.y0!);
+    if (minPadding < totalPaddingInColumn) {
+      // Here we are calculating the min of default and calculated padding, we will not increase the padding
+      // in any scenario.
+      padding = Math.min(padding, minPadding / (column.length - 1));
+    }
+  });
+  sankey.nodePadding(padding);
+}
+
+function idFromNumberOrSNode(node: SNode | number): number {
+  if (typeof node === 'number') {
+    return node;
+  }
+  return node.nodeId as number;
+}
+
+/**
+ * Duplicates the supplied chart data so that we do not alter the original.
+ * @param data The data to duplicate.
+ * @returns The duplicated data.
+ */
+function duplicateData(data: ISankeyChartData): ISankeyChartData {
+  return {
+    nodes: data.nodes.map(
+      // NOTE: We are not duplicating using `...node` because we want to omit the `sourceLinks` and `targetLinks`
+      (node: SNode): SNode => ({
+        nodeId: node.nodeId,
+        name: node.name,
+        color: node.color,
+        borderColor: node.borderColor,
+      }),
+    ),
+    links: data.links.map(
+      (link: SLink): SLink => ({
+        source: idFromNumberOrSNode(link.source),
+        target: idFromNumberOrSNode(link.target),
+        value: link.value,
+      }),
+    ),
+  };
+}
+
+function valuesOfNodes(nodes: SNode[]): NodeValues {
+  const result: NodeValues = {};
+  nodes.forEach((node: SNode) => {
+    result[node.nodeId as number] = node.value!;
+  });
+  return result;
+}
+
+function valuesOfLinks(links: SLink[]): LinkValues {
+  const result: LinkValues = {};
```

Comment:
>   [](http://example.com/codeflow?start=0&length=2)

ToDo: Not a PR blocker.
There are a lot of calculations happening here. Calls for adding UTs to test the individual logic.